common/djangoapps/request_cache/__init__.py

```diff
@@ -26,3 +34,38 @@ def get_request():
     Return the current request.
     """
     return middleware.RequestCache.get_current_request()
+
+
+def get_request_or_stub():
+    """
+    Return the current request or a stub request.
+
+    If called outside the context of a request, construct a fake
+    request that can be used to build an absolute URI.
+
+    This is useful in cases where we need to pass in a request object
+    but don't have an active request (for example, in test cases).
```

Comment:
DRF v3.1 added the requirement that serializers with a "hyperlink" field are instantiated with an HTTP request.  In many parts of the system, we are using serializers within Python APIs (`api.py` convention), and the test cases frequently invoke these APIs outside the context of a request.

I wanted to avoid changing the interfaces to the Python APIs (e.g. not passing in a request to particular API calls).  Changing the interfaces would leak an implementation detail (that we're using DRF serializers) and require updating every instance of the API call.

Instead, I decided to have the APIs that need to retrieve the request from the request cache (stored by middleware).

This works for invocations of the APIs in production code (which are all invoked in the context of a request), but fail with an `AssertionError` from DRF in test cases that invoke the API outside the context of a request.

My reasoning for having this function return a stub is:

1) It's better to log an error and return a stub request that DRF can use to construct URLs than to raise a 500 error in production.

2) It's better to make this change in one place rather than updating loads of tests to patch `get_request()`.  (There are a ton of these).